org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java

```diff
@@ -273,9 +273,9 @@ public static void push(final MethodVisitor mv, final int value) {
 	 */
 	public static ClassReader classReaderFor(final byte[] b) {
 		final int originalVersion = getMajorVersion(b);
-		if (originalVersion == Opcodes.V15 + 1) {
+		if (originalVersion == Opcodes.V16 + 1) {
```

Comment:
If you prefer we can remove support for Java 17. But in any case I would not downgrade Java 16 files otherwise we're not actually using (and testing) ASM for Java 16 files.

Reply:
Well strictly speaking we don't need to test that ASM works with `Opcodes.V16`
only that JaCoCo works for Java 16 class files with or without downgrade 😜 
But
```
if (originalVersion == Opcodes.V16) {
	setMajorVersion(Opcodes.V16, b);
```
also sounds good to me.

Reply:
Not sure whether I understand this correctly: So you're suggesting that we don't do any downgrade at the moment? This means we would also remove the corresponding tests for now?

Reply:
As far as I understand this is exactly what you were suggesting in https://github.com/jacoco/jacoco/pull/1094#discussion_r496037888

> I would not downgrade Java 16

except that we don't need to remove any tests, isn't it?

Reply:
If I understand correctly my original proposal is:

- Java 16 class file: leave as is
- Java 17 class file: downgrade to 16

I simply did it this because I was lazy and

1. We actually use ASM for Java 16
2. We keep the downgrade infrastructure and tests in place
3. (we're prepared for java 17...)

Reply:
My proposal - is all except

> 3. (we're prepared for java 17...)

```diff --git c/org.jacoco.build/licenses/asm-8.0.1.html w/org.jacoco.build/licenses/asm-9.0.html
rename from org.jacoco.build/licenses/asm-8.0.1.html
rename to org.jacoco.build/licenses/asm-9.0.html
index 1ba58e51..7e6d1d7b 100644
--- c/org.jacoco.build/licenses/asm-8.0.1.html
+++ w/org.jacoco.build/licenses/asm-9.0.html
@@ -1,7 +1,7 @@
 <h4>ASM</h4>
 
 <p>
-  <a href="http://asm.objectweb.org/">ASM 7.2</a> is subject to the terms and
+  <a href="http://asm.objectweb.org/">ASM 9.0</a> is subject to the terms and
   conditions of the following license:
 </p>
 
diff --git c/org.jacoco.build/pom.xml w/org.jacoco.build/pom.xml
index 078b4e73..89acca23 100644
--- c/org.jacoco.build/pom.xml
+++ w/org.jacoco.build/pom.xml
@@ -139,7 +139,7 @@
     <argLine>${jvm.args}</argLine>
 
     <!-- Dependencies versions -->
-    <asm.version>8.0.1</asm.version>
+    <asm.version>9.0</asm.version>
     <ant.version>1.7.1</ant.version>
     <args4j.version>2.0.28</args4j.version>
     <junit.version>4.13</junit.version>
diff --git c/org.jacoco.core.test/src/org/jacoco/core/analysis/AnalyzerTest.java w/org.jacoco.core.test/src/org/jacoco/core/analysis/AnalyzerTest.java
index 120ddb92..33d1ea5a 100644
--- c/org.jacoco.core.test/src/org/jacoco/core/analysis/AnalyzerTest.java
+++ w/org.jacoco.core.test/src/org/jacoco/core/analysis/AnalyzerTest.java
@@ -107,7 +107,7 @@ public class AnalyzerTest {
 	@Test
 	public void should_not_modify_class_bytes_to_support_next_version()
 			throws Exception {
-		final byte[] originalBytes = createClass(Opcodes.V15 + 1);
+		final byte[] originalBytes = createClass(Opcodes.V16);
 		final byte[] bytes = new byte[originalBytes.length];
 		System.arraycopy(originalBytes, 0, bytes, 0, originalBytes.length);
 		final long expectedClassId = CRC64.classId(bytes);
@@ -130,7 +130,7 @@ public class AnalyzerTest {
 	 */
 	@Test
 	public void analyzeClass_should_throw_exception_for_unsupported_class_file_version() {
-		final byte[] bytes = createClass(Opcodes.V15 + 2);
+		final byte[] bytes = createClass(Opcodes.V16 + 1);
 		try {
 			analyzer.analyzeClass(bytes, "UnsupportedVersion");
 			fail("exception expected");
@@ -217,7 +217,7 @@ public class AnalyzerTest {
 	 */
 	@Test
 	public void analyzeAll_should_throw_exception_for_unsupported_class_file_version() {
-		final byte[] bytes = createClass(Opcodes.V15 + 2);
+		final byte[] bytes = createClass(Opcodes.V16 + 1);
 		try {
 			analyzer.analyzeAll(new ByteArrayInputStream(bytes),
 					"UnsupportedVersion");
diff --git c/org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java w/org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java
index f1d21db2..3afa8103 100644
--- c/org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java
+++ w/org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java
@@ -30,6 +30,7 @@ import static org.objectweb.asm.Opcodes.V12;
 import static org.objectweb.asm.Opcodes.V13;
 import static org.objectweb.asm.Opcodes.V14;
 import static org.objectweb.asm.Opcodes.V15;
+import static org.objectweb.asm.Opcodes.V16;
 import static org.objectweb.asm.Opcodes.V1_1;
 import static org.objectweb.asm.Opcodes.V1_2;
 import static org.objectweb.asm.Opcodes.V1_3;
@@ -135,7 +136,7 @@ public class ClassFileVersionsTest {
 
 	@Test
 	public void test_16() throws IOException {
-		testVersion(V15 + 1, true);
+		testVersion(V16, true);
 	}
 
 	private void testVersion(int version, boolean frames) throws IOException {
diff --git c/org.jacoco.core.test/src/org/jacoco/core/instr/InstrumenterTest.java w/org.jacoco.core.test/src/org/jacoco/core/instr/InstrumenterTest.java
index 5d29429a..b7500cb1 100644
--- c/org.jacoco.core.test/src/org/jacoco/core/instr/InstrumenterTest.java
+++ w/org.jacoco.core.test/src/org/jacoco/core/instr/InstrumenterTest.java
@@ -99,7 +99,7 @@ public class InstrumenterTest {
 	@Test
 	public void should_not_modify_class_bytes_to_support_next_version()
 			throws Exception {
-		final byte[] originalBytes = createClass(Opcodes.V15 + 1);
+		final byte[] originalBytes = createClass(Opcodes.V16);
 		final byte[] bytes = new byte[originalBytes.length];
 		System.arraycopy(originalBytes, 0, bytes, 0, originalBytes.length);
 		final long expectedClassId = CRC64.classId(bytes);
@@ -122,7 +122,7 @@ public class InstrumenterTest {
 	 */
 	@Test
 	public void instrument_should_throw_exception_for_unsupported_class_file_version() {
-		final byte[] bytes = createClass(Opcodes.V15 + 2);
+		final byte[] bytes = createClass(Opcodes.V16 + 1);
 		try {
 			instrumenter.instrument(bytes, "UnsupportedVersion");
 			fail("exception expected");
@@ -224,7 +224,7 @@ public class InstrumenterTest {
 	 */
 	@Test
 	public void instrumentAll_should_throw_exception_for_unsupported_class_file_version() {
-		final byte[] bytes = createClass(Opcodes.V15 + 2);
+		final byte[] bytes = createClass(Opcodes.V16 + 1);
 		try {
 			instrumenter.instrumentAll(new ByteArrayInputStream(bytes),
 					new ByteArrayOutputStream(), "UnsupportedVersion");
diff --git c/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java w/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
index b6ddb293..a66b5ebd 100644
--- c/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
+++ w/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
@@ -53,7 +53,7 @@ public class InstrSupportTest {
 			public void visit(final int version, final int access,
 					final String name, final String signature,
 					final String superName, final String[] interfaces) {
-				assertEquals(Opcodes.V15 + 1, version);
+				assertEquals(Opcodes.V16, version);
 			}
 		}, 0);
 
@@ -62,7 +62,7 @@ public class InstrSupportTest {
 
 	private static byte[] createJava16Class() {
 		final ClassWriter cw = new ClassWriter(0);
-		cw.visit(Opcodes.V15 + 1, 0, "Foo", null, "java/lang/Object", null);
+		cw.visit(Opcodes.V16, 0, "Foo", null, "java/lang/Object", null);
 		cw.visitEnd();
 		return cw.toByteArray();
 	}
@@ -127,7 +127,7 @@ public class InstrSupportTest {
 		assertTrue(InstrSupport.needsFrames(Opcodes.V13));
 		assertTrue(InstrSupport.needsFrames(Opcodes.V14));
 		assertTrue(InstrSupport.needsFrames(Opcodes.V15));
-		assertTrue(InstrSupport.needsFrames(Opcodes.V15 + 1));
+		assertTrue(InstrSupport.needsFrames(Opcodes.V16));
 
 		assertTrue(InstrSupport.needsFrames(0x0100));
 	}
diff --git c/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java w/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
index bf91e6d0..a4de1f22 100644
--- c/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ w/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -273,9 +273,9 @@ public final class InstrSupport {
 	 */
 	public static ClassReader classReaderFor(final byte[] b) {
 		final int originalVersion = getMajorVersion(b);
-		if (originalVersion == Opcodes.V15 + 1) {
+		if (originalVersion == Opcodes.V16) {
 			// temporarily downgrade version to bypass check in ASM
-			setMajorVersion(Opcodes.V15, b);
+			setMajorVersion(Opcodes.V16, b);
 		}
 		final ClassReader classReader = new ClassReader(b);
 		setMajorVersion(originalVersion, b);
diff --git c/org.jacoco.doc/docroot/doc/changes.html w/org.jacoco.doc/docroot/doc/changes.html
index 67769e46..8558948b 100644
--- c/org.jacoco.doc/docroot/doc/changes.html
+++ w/org.jacoco.doc/docroot/doc/changes.html
@@ -20,6 +20,19 @@
 
 <h2>Snapshot Build @qualified.bundle.version@ (@build.date@)</h2>
 
+<h3>New Features</h3>
+<ul>
+  <li>JaCoCo now officially supports Java 15
+      (GitHub <a href="https://github.com/jacoco/jacoco/issues/1094">#1094</a>).</li>
+</ul>
+
+<h3>Non-functional Changes</h3>
+<ul>
+  <li>JaCoCo now depends on ASM 9.0
+      (GitHub <a href="https://github.com/jacoco/jacoco/issues/1094">#1094</a>).</li>
+</ul>
+
+
 <h2>Release 0.8.6 (2020/09/15)</h2>
 
 <h3>New Features</h3>
```

Reply:
Ok, this means no downgrading at all ATM?

Reply:
> Ok, this means no downgrading at all ATM?

yup, still infrastructure and tests in place

Reply:
Can you please directly push the changes that you have in mind?

Reply:
Thanks!

Reply:
My pleasure.